src/Parallel/Main.hpp

```diff
@@ -122,6 +124,65 @@ class Main : public CBase_Main<Metavariables> {
       phase_change_decision_data_;
 };
 
+namespace detail {
+
+// Charm++ AtSync effectively requires an additional global sync to the
+// quiescence detection we do for switching phases. However, AtSync only needs
+// to be called for one array to trigger the sync-based load balancing, so the
+// AtSyncIndicator is a silly hack to have a centralized indication to start
+// load-balancing. It participates in the `AtSync` barrier, but is not
+// migratable, and should only be constructed by the `Main` chare on the same
+// processor as the `Main` chare. When load-balancing occurs, main invokes the
+// member function `IndicateAtSync()`, and when `ResumeFromSync()` is called
+// from charm, `AtSyncIndicator` simply passes control back to the Main chare
+// via `execute_next_phase()`.
+template <typename Metavariables>
+class AtSyncIndicator : public CBase_AtSyncIndicator<Metavariables> {
```

Comment:
- Do I understand correctly that this is the only chare that sets `usesAtSync` to true and also the only chare that calls `AtSync()`, and thus it essentially just triggers load-balancing right away?
- Are you currently planning to do load-balancing always in a separate phase, i.e. at global synchronization points, or might it also be possible to load-balance with the "periodic" mode that charm++ offers?

Reply:
Yes, the first statement is correct.
Yes, we currently plan to only load-balance at global sync points (at least for this first iteration). The reason for this is the registration/deregistration problems. That process would need a lot of changes to support periodic balancing, with quite a bit of bookkeeping to collect and forward messages that aren't synched-up between points in the evolution in the origin and destination node for a migrated element.
This was something that we discussed when we started testing load-balancing and I have a rough design sketch for the algorithm necessary to do periodic balancing, but we made the decision that it would be enough of a technical complication that it should be tackled separately and only if we can ascertain that the global syncs are causing a bottleneck.

That said, if you aren't using observers or anything else that you register with on a node-by-node basis in spectre code (e.g. interpolators), you can just leave out the `LoadBalancing` phase, and include periodic balancing arguments to the executable and it will just work. If you do that while trying to use observers though, it will quickly error with registration problems.

Reply:
I think (correct me if I'm misremembering!) that dynamic insertion and deletion of array elements like what AMR will do also isn't currently support by load-balancing unless you can guarantee that it won't insert/delete during the LB. The only real way to do that is a global sync. We could relax the global sync a bit to being a sync only over array components (we won't be migrating groups/nodegroups) so that we don't have to wait on IO to finish before we can load balance. But that's an optimization way down the line, I think

Reply:
Yes, the interaction with AMR is a very good point, and the constraint of not inserting/deleting is consistent with my understanding of the charm++ requirements.